src/chkreg/chkreg.c

```diff
@@ -0,0 +1,447 @@
+
+    /* Only stop the module when it is not currently used by other modules. */
+
+    status = sceKernelWaitSema(g_semaId, 1, &timeout); // 0x0x0000035C
+    if (status == SCE_ERROR_OK) // 0x0x00000364
+    {
+        /* Module no longer in use. Stop it. */
+        sceKernelDeleteSema(g_semaId); // 0x00000380
+
+        return SCE_KERNEL_STOP_SUCCESS;
+    }
+
+    /* Module still in use, don't stop it yet. */
+
+    return SCE_KERNEL_STOP_FAIL;
+}
+
+// Subroutine sceChkreg_driver_54495B19 - Address 0x00000390
+s32 sceChkregCheckRegion(u32 umdMediaType, u32 regionId)
+{
+    s32 status1;
+    s32 status2;
+
+    status1 = SCE_ERROR_SEMAPHORE; // 0x000003BC & 0x000003C8
+
+    /* Allow only one access at a time. */
+    status2 = sceKernelWaitSema(g_semaId, 1, NULL); // 0x000003C4
+    if (status2 == SCE_ERROR_OK)
+    {
+        if (g_isUMDRegionCodesObtained || (status1 = _sceChkregLookupUMDRegionCodeInfo()) == SCE_ERROR_OK) // 0x000003D8 & 0x000003E8
+        {
+            status1 = _sceChkregCheckRegion(0x80000000, umdMediaType | regionId); // 0x000003F4
+        }
+
+        /* Release acquired sema resource. */
+        status2 = sceKernelSignalSema(g_semaId, 1); // 0x00000404
+        if (status2 != SCE_ERROR_OK)
+        {
+            status1 = SCE_ERROR_SEMAPHORE;
+        }
+    }
+
+    return status1;
+}
+
+// Subroutine sceChkreg_driver_59F8491D - Address 0x00000438
+s32 sceChkregGetPsCode(ScePsCode *pPsCode)
+{
+    s32 status1;
+    s32 status2;
+
+    status1 = SCE_ERROR_SEMAPHORE;
+
+    /* Allow only one access at a time. */
+    status2 = sceKernelWaitSema(g_semaId, 1, NULL);
+    if (status2 == SCE_ERROR_OK)
+    {
+        if ((g_isConsoleIdCertificateObtained || (status1 = _sceChkregLookupConsoleIdCertificate()) == SCE_ERROR_OK)
+            && (status1 = _sceChkregVerifyConsoleIdCertificate()) == SCE_ERROR_OK)
+        {
+            pPsCode->companyCode = g_ConsoleIdCertificate.consoleId.companyCode;
+            pPsCode->productCode = g_ConsoleIdCertificate.consoleId.productCode;
+            pPsCode->productSubCode = g_ConsoleIdCertificate.consoleId.productSubCode;
+            pPsCode->factoryCode = g_ConsoleIdCertificate.consoleId.factoryCode;
+        }
+
+        /* Release acquired sema resource. */
+        status2 = sceKernelSignalSema(g_semaId, 1);
+        if (status2 != SCE_ERROR_OK)
+        {
+            status1 = SCE_ERROR_SEMAPHORE;
+        }
+    }
+
+    return status1;
+}
+
+// Subroutine sceChkreg_driver_9C6E1D34 - Address 0x0000051C
+s32 sceChkreg_driver_9C6E1D34(const u8 *arg0, u8 *arg1)
+{
+    s32 status1;
+    s32 status2;
+
+    status1 = SCE_ERROR_SEMAPHORE;
+
+    /* Allow only one access at a time. */
+    status2 = sceKernelWaitSema(g_semaId, 1, NULL); // 0x00000554
+    if (status2 == SCE_ERROR_OK) // 0x0000055C
+    {
+        u8 *pWorkBuffer = g_pWorkBuffer;
+
+        ((u32 *)pWorkBuffer)[0] = 0x34; // 0x00000570 -- TODO: Length of specific data to hash?
```

Comment:
I think it's indeed the length of the data to hash — better use a structure (with only 2 elements and unspecified size)